text/0000-structural-records.md

```diff
@@ -0,0 +1,1262 @@
+in parenthesis. To avoid ambiguities with the `{ x }` syntax that is already
+allowed, we impose the same rule for one-field structural records. To create
+such a record, you must write `{ x, }` or `{ x: <value>, }`. This difference
+is also imposed to maintain a sense of consistency between the positional
+and the non-positional structural variant.
+
+### Struct update syntax
+
+Nominal structs support what is referred to as the *"struct update syntax"*,
+otherwise known as functional record update (FRU). For example, you can write:
+
+```rust
+struct Color {
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+let yellow = Color { red: 0, green: 255, blue: 255 };
+
+let white = Color { red: 255, ..yellow };
+```
+
+This also works for structural records, so you can write:
+
+```rust
+let yellow = { red: 0, green: 255, blue: 255 };
+
+let white = { red: 255, ..yellow };
+```
+
+To match the behaviour of FRU for nominal structs, we impose a restriction
+that the fields mentioned before `..yellow` must all exist in `yellow` and
+have the same types. This means that we cannot write:
+
+```rust
+let white_50_opacity = { alpha: 0.5, red: 255, ..yellow };
+```
+
+However, there is no fundamental reason why we could not allow this,
+but to start off more conservative and to be uniform in behaviour,
+we don't allow this for the time being.
+
+### The type of a structural record
+
+You might be wondering what the type of `my_record` that we've been using
+thus far is. Because this is structural typing, the fields are significant,
+so the type of the record is simply:
+
+```rust
+type TheRecord = { alpha: u8, beta: bool, gamma: &'static str };
+
+let my_record: TheRecord = { alpha: 42u8, beta: true, gamma: "Dancing Ferris" };
+```
+
+Notice how this matches the way we defined `MyRecordType` is we remove
+the prefix `struct MyRecordType`.
+
+The order in which we've put `alpha`, `beta`, and `gamma` here does not matter.
+We could have also written:
+
+```rust
+type TheRecord = { beta: bool, alpha: u8, gamma: &'static str };
+```
+
+As long as the type is a *permutation* of each pair of field name and field type,
+the type is the same. This also means that we can write: 
+
+```rust
+let my_record: TheRecord = { alpha: 42u8, gamma: "Dancing Ferris", beta: true };
+```
+
+### Implemented traits
+
+With respect to trait implementations, because the type is structural,
+and because there may be an unbound number of fields that can all be arbitrary
+identifiers, there's no way to define implementations for the usual traits
+in the language itself.
+
+[tuples]: https://doc.rust-lang.org/std/primitive.tuple.html
+
+Instead, the compiler will automatically provide trait implementations for the
+standard traits that are implemented for [tuples]. These traits are: `Clone`,
+`Copy`, `PartialEq`, `Eq`, `PartialOrd`, `Ord`, `Debug`, `Default`, and `Hash`.
+Each of these traits will only be implemented if all the field types of a struct
+implements the trait.
```

Comment:
That's... quite a bit of magic.

Reply:
I agree; this is noted in [the drawbacks](https://github.com/Centril/rfcs/blob/rfc/structural-records/text/0000-structural-records.md#auto-implementing-traits-is-a-magical-hack).

Reply:
Ideas for possible magic reduction discussed below in https://github.com/rust-lang/rfcs/pull/2584#issuecomment-435556892.